src/evolve/downloader.go

```diff
@@ -0,0 +1,216 @@
+					Error("Failed to download contract source transaction")
+				contractSrc.SrcTxId = *srcId
+				contractSrc.Src.Set("error")
+				goto end
+			} 
+
+			contractSrc, err = self.loadContractSrcMetadata(tx)
+			if err != nil {
+				self.Log.WithError(err).
+					WithField("srcId", srcId).
+					Error("Failed to load contract source metadata")
+				contractSrc.SrcTxId = *srcId
+				contractSrc.Src.Set("error")
+			}
+
+			end:
+				self.Output <- contractSrc
+		})
+	}
+
+	return nil
+}
+
+
+func (self *Downloader) downloadContractSrcTx(downloadTxFunc downloadTxFunc, srcId string) (out *arweave.Transaction, err error) {
+	err = task.NewRetry().
+		WithContext(self.Ctx).
+		WithMaxElapsedTime(0).
+		WithMaxInterval(self.Config.Evolve.SourceTransactionDownloadMaxInterval).
+		WithAcceptableDuration(self.Config.Evolve.SourceTransactionDownloadMaxInterval * 3).
+		WithOnError(func(err error, isDurationAcceptable bool) error {
+			// Permanent errors
+			if errors.Is(err, context.Canceled) && self.IsStopping.Load() {
+				return backoff.Permanent(err)
+			}
+
+			if errors.Is(err, arweave.ErrNotFound) {
+				return backoff.Permanent(err)
+			}
+
+			if errors.Is(err, arweave.ErrOverspend) {
+				return backoff.Permanent(err)
+			}
+
+			// Errors fo retry
+			self.Log.WithError(err).WithField("srcId", srcId).Warn("Failed to download contract source transaction, retrying after timeout")
+			
+			if errors.Is(err, arweave.ErrPending) {
+				time.Sleep(time.Second)
+				return err
+			}
+
+			if !isDurationAcceptable {
+				self.client.Reset()
+			}
+
+			return err
+		}).
+		Run(func() error {
+			out, err = downloadTxFunc(self.Ctx, srcId)
+			if err != nil {
+				self.Log.WithField("txId", srcId).Error("Failed to download contract source transaction")
+				return err
+			}
+			return err
+		})
+	if err != nil {
+		self.Log.WithError(err).WithField("txId", srcId).Error("Failed to download contract source transaction, giving up")
+		return
+	}
+	
+	return
+}
+
+func (self *Downloader) loadContractSrcMetadata(srcTx *arweave.Transaction) (out *model.ContractSource, err error) {
+			out = model.NewContractSource()
+
+			out.SrcTxId = srcTx.ID.Base64()
+
+			srcContentType, ok := srcTx.GetTag(smartweave.TagContentType)
+			if !ok {
+				err = errors.New("contract source content type is not set")
+				return
+			}
+		
+			if !slices.Contains(self.Config.Contract.LoaderSupportedContentTypes, srcContentType) {
+				err = errors.New("unsupported contract source content type")
+				return
+			}
+		
+			err = out.SrcContentType.Set(srcContentType)
+			if err != nil {
+				err = errors.New("could not set contract source type")
+				return
+			}
+
+			owner, err := warp.GetWalletAddress(srcTx)
+			if err != nil {
+				self.Log.WithError(err).Error("could not get wallet address")
+				return
+			}
+		
+			err = out.Owner.Set(owner)
+			if err != nil {
+				err = errors.New("could not set owner")
+				return
+			}
+
+			err = out.SrcTx.Set(srcTx)
```

Comment:
i'm not sure about this one..  `srcTx` is a correct `arweave.Transaction` struct but it looked like binary when I logged it to the console :)